src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -6744,6 +6745,198 @@ def merge(
+                "Snowpark pandas merge_asof method only supports directions 'forward' and 'backward'"
+            )
+
+        left_frame = self._modin_frame
+        right_frame = right._modin_frame
+        # Project to Snowpark DataFrame in order to perform ASOF Join
+        left_snowpark_df = (
+            left_frame.ordered_dataframe.to_projected_snowpark_dataframe()
+        )
+        right_snowpark_df = (
+            right_frame.ordered_dataframe.to_projected_snowpark_dataframe()
+        )
+        if on:
+            on_quoted_identifier = quote_name_without_upper_casing(on)
+            lhs_expr = left_snowpark_df[on_quoted_identifier]
+            rhs_expr = right_snowpark_df[on_quoted_identifier]
+            order_by_expr = [
+                left_snowpark_df[on_quoted_identifier],
+                right_snowpark_df[on_quoted_identifier],
+            ]
+        else:
+            assert left_on and right_on
+            left_on_quoted_identifier = quote_name_without_upper_casing(left_on)
+            right_on_quoted_identifier = quote_name_without_upper_casing(right_on)
+
+            lhs_expr = left_snowpark_df[left_on_quoted_identifier]
+            rhs_expr = right_snowpark_df[right_on_quoted_identifier]
+            order_by_expr = [
+                left_snowpark_df[left_on_quoted_identifier],
+                right_snowpark_df[right_on_quoted_identifier],
+            ]
+
+        if direction == "backward":
+            match_condition = (
+                (lhs_expr >= rhs_expr) if allow_exact_matches else (lhs_expr > rhs_expr)
+            )
+        else:
+            match_condition = (
+                (lhs_expr <= rhs_expr) if allow_exact_matches else (lhs_expr < rhs_expr)
+            )
+
+        joined_snowpark_df = left_snowpark_df.join(
+            right=right_snowpark_df,
+            match_condition=match_condition,
+            how="asof",
+        ).order_by(order_by_expr)
+
+        if on:
+            # Output pandas labels consists of all data pandas labels in left_frame and
+            # all data pandas labels except the "on" pandas label in the right frame
+            right_frame_on_index = right_frame.data_column_pandas_labels.index(on)
+            data_column_pandas_labels = (
+                left_frame.data_column_pandas_labels
+                + right_frame.data_column_pandas_labels[0:right_frame_on_index]
+                + right_frame.data_column_pandas_labels[(right_frame_on_index + 1) :]
+            )
+        else:
+            # Output pandas labels consists of all data pandas labels in both frames
+            assert left_on and right_on
+            data_column_pandas_labels = (
+                left_frame.data_column_pandas_labels
+                + right_frame.data_column_pandas_labels
+            )
+
+        left_frame_num_valid_cols = len(left_frame.data_column_pandas_labels)
+        right_frame_num_valid_cols = len(right_frame.data_column_pandas_labels)
```

Comment:
Can we move this into the join_utils file?

Reply:
Which lines do you think should be moved into `join_utils`? I think most of this code is specific to asof join so isn't general purpose enough to warrant a function in `join_utils`, but I can make it one if you think it's better.

Reply:
Ah I was thinking more perhaps adding the whole do the join + map to new labels to a function (either the general join function, or a specific asof join function) in join_utils.py so that way you don't have to do the mapping here. (Also potentially check if theres code in join_utils.py that you can reuse to get the mapping). Basically, have some function that returns a JoinOrAlignResult (as would happen if you did a traditional join).